### PR TITLE
Buffer overwritten caused PK migration failure

### LINKED ISSUE #1775: Primary key not defined for field '_id' in existing Realm file. Add @PrimaryKey.
This is my RealmConfiguration：
stableCongifuration = new RealmConfiguration.Builder(context).deleteRealmIfMigrationNeeded().name("stable.realm").build();
 
when id call Realm.getInstance(stableCongifuration); I got this exception:

11-14 06:27:18.829 24386-24386/? E/AndroidRuntime:  Caused by: io.realm.exceptions.RealmMigrationNeededException: Primary key not defined for field '_id' in existing Realm file. Add @PrimaryKey.
11-14 06:27:18.829 24386-24386/? E/AndroidRuntime:     at io.realm.JokeRealmProxy.validateTable(JokeRealmProxy.java:443)
11-14 06:27:18.829 24386-24386/? E/AndroidRuntime:     at io.realm.DefaultRealmModuleMediator.validateTable(DefaultRealmModuleMediator.java:78)
11-14 06:27:18.829 24386-24386/? E/AndroidRuntime:     at io.realm.Realm.initializeRealm(Realm.java:342)
11-14 06:27:18.829 24386-24386/? E/AndroidRuntime:     at io.realm.Realm.createAndValidate(Realm.java:310)
11-14 06:27:18.829 24386-24386/? E/AndroidRuntime:     at io.realm.Realm.create(Realm.java:253)
11-14 06:27:18.829 24386-24386/? E/AndroidRuntime:     at io.realm.Realm.getInstance(Realm.java:217)
11-14 06:27:18.829 24386-24386/? E/AndroidRuntime:     at com.qiuduanribao.joke.utils.RealmUtils.getStableRealm(RealmUtils.java:32)

and the "_id" field have add @PrimaryKey

```
public class User extends RealmObject {
    @PrimaryKey
    private Long _id;
    private RealmList<Auth> auths;
    private String token;
    private String name;
    private boolean isMan;
    private String avatar;
}
```

Why ? How can i fix this ?

### PR DESCRIPTION
**Pull request by beeender**
Close #1775
A String or Binary pointer retrieved from Realm is invalidated when
set_string called.

### CODE REVIEW

**Comment by beeender** [comment]
@realm/java @rrrlasse

**Comment by kneth** [comment]
:+1:

**Comment by cmelchior** [comment]
:+1:

**Comment by emanuelez** [comment]
:+1:

**Comment by dalinaum** [comment]
:+1:

File: realm/realm-jni/src/io_realm_internal_table.cpp [hunk 1]
```diff
@@ -1586,7 +1594,7 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_Table_nativeSetPrimaryKey(
     try {
         Table* table = TBL(nativeTablePtr);
         Table* pk_table = TBL(nativePrivateKeyTablePtr);
-        const char* table_name = table->get_name().substr(6).data(); // Remove "class_" prefix
+        const std::string table_name(table->get_name().substr(6)); // Remove "class_" prefix
```
Comment (zaki50) [inline realm/realm-jni/src/io_realm_internal_table.cpp]: Is this copy needed? I think `substr()` always copy the data.

Comment (beeender) [inline realm/realm-jni/src/io_realm_internal_table.cpp]: `substr` of `std::string` does copy. `StringData`'s doesn't.

Comment (zaki50) [inline realm/realm-jni/src/io_realm_internal_table.cpp]: Oh, I understand. It's confusing...

**Pull request merged**